src/core/thread/mesh_forwarder.cpp

```diff
@@ -488,19 +494,33 @@ bool MeshForwarder::IsDirectTxQueueOverMaxFrameThreshold(void) const
     return (frameCount > OPENTHREAD_CONFIG_MAX_FRAMES_IN_DIRECT_TX_QUEUE);
 }
 
+void MeshForwarder::AgedMessageDropped(Message *aMessage, void *aContext)
+{
+    AgedMessageDroppedContext *context = static_cast<AgedMessageDroppedContext *>(aContext);
+
+    if (context->mMessage == aMessage)
+    {
+        *context->mWasDropped = true;
+    }
+}
+
 void MeshForwarder::ApplyDirectTxQueueLimit(Message &aMessage)
 {
+    bool                      wasDropped = false;
+    AgedMessageDroppedContext context    = {&aMessage, &wasDropped};
+
     VerifyOrExit(aMessage.IsDirectTransmission());
     VerifyOrExit(IsDirectTxQueueOverMaxFrameThreshold());
 
 #if OPENTHREAD_CONFIG_DELAY_AWARE_QUEUE_MANAGEMENT_ENABLE
-    if (RemoveAgedMessages() == kErrorNone)
+    if (RemoveAgedMessages(AgedMessageDropped, &context) == kErrorNone)
     {
         VerifyOrExit(IsDirectTxQueueOverMaxFrameThreshold());
     }
 #endif
 
     LogMessage(kMessageFullQueueDrop, aMessage);
+    VerifyOrExit(wasDropped == false);
```

Comment:
Thanks @pkwiek for finding and fixing this.

Checking the code, I see the following code path:
- From `MeshForwarder::SendMessage()` (we have ftd/mtd flavors but both do the same related to this):
- We enqueue the message
- And set its timestamp to now `mSendQueue.Enqueue(aMessage)`
- And then at the end of the same method, we call `ApplyDirectTxQueueLimit()`
- Which then may call `RemoveAgedMessages()` 
   - Which may remove the new message if it is aged more than 0.5 seconds
- And all of this is from the same call path

I agree that it can be possible for this to happen, but I think this should be very unlikely. We need the call flow to be interrupted (e.g. by some other task/ISR) to block execution for more than 0.5 second before we get back, for the new message in the queue to be considered aged and removed.

Does this sound correct to you? Do you happen to know any other code paths where this can happen?

Reply:
It looks exactly like the case we are observing. In some of our tests the host works under a heavy load, so the scenario becomes more probable.
I am not aware of any other code paths.

Reply:
Great. Thanks.